Implement canImport test for submodule

ApolloZhu opened the pull request:
<!-- What's in this pull request? -->
This PR extends `canImport` to check for submodule availability.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Forums Thread: https://forums.swift.org/t/41196

### Motivation

Currently, `canImport` only supports checking for top-level module availability while many have submodules with different availability. For example, while `CIFilter` was introduced in iOS 5, `CIFilterBuiltins` is only available on iOS 13+. Therefore, one might want to write code like this:

```swift
#if canImport(CoreImage.CIFilterBuiltins)
    ^ Unexpected platform condition argument: expected identifier

import CoreImage.CIFilterBuiltins
...
#endif
```

### Changes needed for *The Swift Programming Language* Reference

This PR would mean:

```diff
+ module-name → import-path
- module-name → identifier
```

For housekeeping, because [`import OpenGL.(GL3, GL3.Ext)`](https://github.com/apple/swift/blob/main/docs/archive/Import.rst) is (likely) never implemented, unless we're planning to have a separate PR extending import to support operators (e.g. `import func SomeModule.+++`) as mentioned in  https://github.com/apple/swift/pull/34094#issuecomment-708709198, we should update:

```diff
+ import-path → identifier | identifier . import-path
- import-path → import-path-identifier | import-path-identifier . import-path
- import-path-identifier → identifier | operator 
```

### Alternatives Considered

While the initial concern could be addressed by introducing `#if available`similiar to how `@available/#available` works but at the top level, clang modules [may specify other requirements](https://clang.llvm.org/docs/Modules.html#requires-declaration), therefore allowing `canImport` to check for submodule availability directly should be a more well-rounded and direct approach.

### Questions/Concerns

- **Is this a bug fix or does it have to go through Swift Evolution?** According to the [original proposal SE-0075](https://github.com/apple/swift-evolution/blob/master/proposals/0075-import-test.md), "`#if canImport(module-name)` tests for module support by name." Are submodule names module names?
- **Current implementation only extends support to clang submodules.** I saw a few `FIXME`s in other module loaders mentioning submodule support is not in place yet. Does this change have to wait for those to be implemented?


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

ApolloZhu commented on test/Parse/ConditionalCompilation/canimport_in_inactive.swift:
```diff
@@ -21,7 +21,7 @@
 
 #if canImport(SomeModule)
-// CHECK: :[[@LINE-1]]:5: error: could not find module 'SomeModule' for target '{{.*}}'; found: i386
+// CHECK: :[[@LINE-1]]:15: error: could not find module 'SomeModule' for target '{{.*}}'; found: i386
```
Now emits diagnostics at the module name instead canImport

```diff
#if canImport(NotExist)
-   ^
+             ^
```

ApolloZhu commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -1709,12 +1709,16 @@ bool ClangImporter::isModuleImported(const clang::Module *M) {
+bool ClangImporter::canImportModule(ImportPath::Module path) {
+  if (path.hasSubmodule()) {
+    // TODO: more efficient way to check for submodule availability?
+    return bool(loadModule(path.front().Loc, path));
```
Is loading the module an acceptable way to check? What would be a better approach?

I tried to iterate through `submodules()` of `clang::Module` directly, but it traps at finding the iterator front. I assume you have to load the module to find its submodules?

CodaFi replied:
Unfortunately, this approach is too expensive and potentially source-breaking. `canImport` checks are frequently run by clients that expect lightning-fast responses from parse-only passes. We need to use the appropriate lookup paths instead. Additionally, loading modules potentially makes those symbols visible.

ApolloZhu replied:
Make sense. I'll try to figure out how to do that.

ApolloZhu commented on test/Parse/ConditionalCompilation/can_import_submodule_missing_requirement.swift:
```diff
@@ -0,0 +1,22 @@
+// RUN: %target-typecheck-verify-swift -I %S/Inputs/can-import-submodule-missing-requirement/ 
+
+#if canImport(A.BMissingRequirement)
+#error("should not can import A.BMissingRequirement")
+#endif
```
I'm thinking about how to make sure `canImport` failed because of missing requirements, not because of miss-spelling, etc. Therefore, I want to write test like:

```swift
#if canImport(A.BMissingRequirement)
#error("should not can import A.BMissingRequirement") // expected-error {{no such module 'A.BMissingRequirement'}}
#else
import A.BMissingRequirement
#endif
```

However, I'm not quite sure how to check for the error `module 'A.BMissingRequirement' requires feature 'missing'` diagnosed at `%S/Inputs/can-import-submodule-missing-requirement/module.modulemap:7:21`.

CodaFi replied:
I wouldn't worry too much about it. External diagnostics are a known shortfall of the verifier's model.

ApolloZhu replied:
Alright then

CodaFi commented on include/swift/AST/Import.h:
```diff
@@ -108,6 +109,14 @@ namespace detail {
+      llvm::interleave(
+          *this, [&](Element elem) { modulePathStr += elem.Item.str(); },
+          [&] { modulePathStr += "."; });
+      return modulePathStr.str();
```
The return of a value derived from a temporary here will cause a use-of-stack-after-return.

CodaFi reviewed (changes requested):
One comment about the implementation and a request: please give this a rebase.

Thank you for tackling this.

ApolloZhu commented:
Thanks for pointing that out! Fixed

CodaFi reviewed (approved):
The implementation looks great.

CodaFi commented:
Please go to swift-evolution and ask if this is a feature or a bug fix. I have confidence that they'll greenlight this, but it's always good to get the community's stamp.

ApolloZhu commented:
Will ask in the forum. Thanks for all the suggestions

ApolloZhu commented:
https://forums.swift.org/t/proposal-or-bug-fix-allow-canimport-to-check-for-submodule-availability/41196

amartini51 commented:
Thanks for calling out the impact on the reference and grammar of "The Swift Programming Language".  I'll follow this PR so I can update the documentation as needed.

The grammar allows operators in an import is probably for the form that allows you to specify a specific declaration to import.  (Git history shows me that grammar rule dates back to the original Swift 1 version of the book.)  For example, something like `import func SomeModule.+++` would import only the function that implements the `+++` operator from `SomeModule`.  However, that form of import doesn't appear to actually work, and I'm unsure whether it was ever actually part of the language's implementation.

ApolloZhu commented:
So far the community response agrees this is a simple bug fix and no further changes need to be made. What shall we do next?

ApolloZhu commented:
I know the community is busy discussing new concurrency proposals, but should I submit a simple Swift Evolution proposal for this change as well?

CodaFi commented:
@ApolloZhu 

Please rebase this when you can and tag me here and we'll get this merged.

ApolloZhu committed:
Extends canImport to check for submodule availability

ApolloZhu commented:
I updated from `main` and squashed all commits into one (since there were too many merge commits in between). Ready for review by @CodaFi.

---

Since our last review of this PR, @nkcsgexi introduced a version check in canImport. Based on their current implementation and a few `.tbd` files I viewed, it seems that submodules don't really have a version. Assuming my observation is correct, we have to consider what to do about:

```swift
#if canImport(A.B.C, _version: 3)
``` 

Two solutions:

- [x] (Current approach) It type checks, validates the version of the top module, and checks if submodule exists.
- [ ] It should NOT type check, and should instead be expressed as:
   ```swift
   #if canImport(A, _version: 3) && canImport(A.B.C)
   ```

nkcsgexi commented:
@ApolloZhu thanks for adding the support for submodules! Though we usually don't have submodule-specific module versions, can we use the top-level module version in the case of sub-modules?

nkcsgexi commented:
@ApolloZhu You are right. Thanks for clarifying!

ApolloZhu commented:
I see this PR is expecting some tests to pass. Who and when can those tests be initiated?

xwu commented:
@swift-ci test

ApolloZhu commented:
1. Fixed merge conflict so this can be merged/tested/reviewed again
2. It would be great if we can remove the `swift-evolution-pending-discussion` label to reflect the confirmation in the earlier comment https://github.com/apple/swift/pull/34094#issuecomment-859162251
3. Please let me know if there are additional things I should do to get this merged
   - We did briefly discuss the potential of implementing code completion for canImport submodule during WWDC, but as there exists no current implementation for canImport top level module completion, I believe that can/should be done in a separate PR

xwu commented:
Pinging @CodaFi as they requested. Tag removed.

xwu commented:
@swift-ci smoke test

CodaFi commented:
Brilliant - still LGTM.  Will merge once the bots come back blue.

CodaFi commented:
@swift-ci smoke test

ApolloZhu commented:
Great, all checks have passed

CodaFi commented:
⛵
